### PR TITLE
feat: Add kfp-metadata-writer charm

### PR DESCRIPTION
**Pull request by orfeas-k**
Introduce `kfp-metadata-writer` charm for upstream KFP V2 [Metadata writer](https://github.com/kubeflow/pipelines/tree/master/backend/metadata_writer) component. Details in the issue below.
Fixes https://github.com/canonical/kfp-operators/issues/348

### CODE REVIEW

File: charms/kfp-metadata-writer/requirements.in [hunk 1]
```diff
@@ -0,0 +1,4 @@
+charmed-kubeflow-chisme >= 0.2.0
```
Comment (DnPlas) [inline charms/kfp-metadata-writer/requirements.in:1]: Please add a comment on why we pin this version

File: charms/kfp-metadata-writer/src/components/pebble_component.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+
+class KfpMetadataWriterPebbleService(PebbleServiceComponent):
+    def __init__(
+        self,
+        *args,
+        namespace_to_watch: str,
+        **kwargs,
+    ):
+        """Pebble service container component in order to configure Pebble layer"""
```
Comment (DnPlas) [inline charms/kfp-metadata-writer/src/components/pebble_component.py:27]:
```suggestion
        """Pebble service container component in order to configure Pebble layer.
        
            Args:
                namespace_to_watch (str): what it is and why it should default to ""
        """
```

File: charms/kfp-metadata-writer/tests/unit/test_operator.py [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+        environment["METADATA_GRPC_SERVICE_SERVICE_HOST"]
+        == harness.charm.grpc_relation.component.get_data()["service"]
+    )
+    assert (
+        environment["METADATA_GRPC_SERVICE_SERVICE_PORT"]
+        == harness.charm.grpc_relation.component.get_data()["port"]
+    )
+
+
+def test_pebble_service_is_replanned_on_config_changed(harness, mocked_lightkube_client):
```
Comment (DnPlas) [inline charms/kfp-metadata-writer/tests/unit/test_operator.py:125]: This test case doesn't seem like checking what you intend. Maybe what you want to do is emit a config changed, and then check that the `replan()` method is called, though it is embedded in the base charm.
On another note, is the expected behaviour of the charm to replan on config changed? I don't see anything that actually triggers that behaviour.

**Review by DnPlas — COMMENTED** [review]
Thanks @orfeas-k, some comments

File: charms/kfp-metadata-writer/metadata.yaml [hunk 4]
```diff
@@ -0,0 +1,33 @@
+name: kfp-metadata-writer
+summary: Reusable end-to-end ML workflows built using the Kubeflow Pipelines SDK
```
[the remark this answers is not in the document]
Comment (orfeas-k) [reply charms/kfp-metadata-writer/metadata.yaml:2]: hmm we have the same summary in all `kfp-` charms, that's why I left it like this

File: charms/kfp-metadata-writer/metadata.yaml [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+website: https://charmhub.io/kfp-metadata-writer
+source: https://github.com/canonical/kfp-operators
+containers:
+  kfp-metadata-writer:
+    resource: oci-image
+resources:
+  oci-image:
+    type: oci-image
+    description: OCI image for KFP Metadata Writer
+    upstream-source: gcr.io/ml-pipeline/metadata-writer:2.0.1
```
Comment (DnPlas) [inline charms/kfp-metadata-writer/metadata.yaml:16]:
```suggestion
    upstream-source: gcr.io/ml-pipeline/metadata-writer:2.0.2
```

**Review by DnPlas — COMMENTED** [review]
Thanks @orfeas-k, some comments about testing and about the workload image version. Other than that, LGTM

File: charms/kfp-metadata-writer/tests/unit/test_operator.py (hunk 3 above)
Comment (orfeas-k) [reply charms/kfp-metadata-writer/tests/unit/test_operator.py:125]: You are right, I was confused from what we tested in other charms. I 'm removing this test as it is pointless.

**Review by DnPlas — APPROVED** [review]
Thanks @orfeas-k !

**Pull request merged**